scipy/spatial/kdtree.py

```diff
@@ -626,17 +477,7 @@ def query_ball_point(self, x, r, p=2., eps=0):
         x = np.asarray(x)
         if x.dtype.kind == 'c':
             raise TypeError("KDTree does not work with complex data")
-        if x.shape[-1] != self.m:
-            raise ValueError("Searching for a %d-dimensional point in a "
-                             "%d-dimensional KDTree" % (x.shape[-1], self.m))
-        if len(x.shape) == 1:
-            return self.__query_ball_point(x, r, p, eps)
-        else:
-            retshape = x.shape[:-1]
-            result = np.empty(retshape, dtype=object)
-            for c in np.ndindex(retshape):
-                result[c] = self.__query_ball_point(x[c], r, p=p, eps=eps)
-            return result
+        return super().query_ball_point(x, r, p, eps)
```

Comment:
Does cKDTree allow complex data? If not, should we assert there instead?

Reply:
`cKDTree` does `ascontiguousarray(x, dtype=np.float64)` everywhere. So you can use complex data, it'll just drop the imaginary components.

Reply:
Yes, and that behavior is not correct. As it is now, the C++ code would actually work correctly with complex data. It is just the Cython wrapper that prevents it.

Reply:
>As it is now, the C++ code would actually work correctly with complex data. It is just the Cython wrapper that prevents it.

It would only work for the special case of `p=2` where we have the property 
```
minkowski_distance(x, y) == sqrt(sqeuclidean(x.real, y.real) + sqeuclidean(x.imag, y.imag))
```
Which works because `abs(z)` for complex `z` is a 2D euclidean norm. But that doesn't hold in general for other values of `p`.

Reply:
I think raising an error for complex data would in any case be better than just silently dropping the complex part and pretending it works correctly.

Reply:
It's not completely silent, NumPy gives you a `ComplexWarning`.